File: day1/go/main.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"slices"
	"strconv"
	"strings"
)

func main() {
	fmt.Printf("DISTANCE: %d\n", partOne("../input.txt"))
	fmt.Printf("SIMILARITY_SCORE: %d\n", partTwo("../input.txt"))
}

func partOne(path string) int {
	firstList, secondList := parse(path)

	slices.Sort(firstList)
	slices.Sort(secondList)

	distances := []float64{}
	for i, first := range firstList {
		second := secondList[i]
		distances = append(distances, math.Abs(float64(first-second)))
	}

	var result int
	for _, d := range distances {
		result += int(d)
	}

	return result
}

func partTwo(path string) int {
	firstList, secondList := parse(path)

	similarityScore := 0
	secondListMap := make(map[int]int)

	for _, second := range secondList {
		secondListMap[second]++
	}

	for _, first := range firstList {
		similarityScore += first * secondListMap[first]
	}

	return similarityScore
}

func parse(path string) ([]int, []int) {
	file, err := os.Open(path)
	handleErr(err)
	defer file.Close()

	scanner := bufio.NewScanner(file)

	firstList := make([]int, 0)
	secondList := make([]int, 0)

	for scanner.Scan() {
		line := scanner.Text()
		split := strings.Fields(line)
		first, err := strconv.Atoi(split[0])
		handleErr(err)
		firstList = append(firstList, first)

		second, err := strconv.Atoi(split[1])
		handleErr(err)
		secondList = append(secondList, second)
	}

	return firstList, secondList
}

func handleErr(e error) {
	if e != nil {
		panic(e)
	}
}

File: day1/rust/src/main.rs
use std::{
    fs::File,
    io::{BufRead, BufReader},
};

fn main() {
    let input_file = File::open("../input.txt").expect("failed to open file");
    let reader = BufReader::new(input_file);

    let mut first_list: Vec<u32> = Vec::new();
    let mut second_list: Vec<u32> = Vec::new();

    for (i, line) in reader.lines().enumerate() {
        let line = line.expect("failed to read line");
        let mut values = line.split_whitespace();
        let first = values
            .next()
            .unwrap_or_else(|| panic!("first value is not provided on line {i}"))
            .parse::<u32>()
            .unwrap();
        first_list.push(first);

        let second = values
            .next()
            .unwrap_or_else(|| panic!("second value is not provided on line {i}"))
            .parse::<u32>()
            .unwrap();
        second_list.push(second);
    }

    first_list.sort_unstable();
    second_list.sort_unstable();

    let mut distances: Vec<u32> = vec![];
    for (i, first) in first_list.iter().enumerate() {
        let second = second_list.get(i).unwrap();
        distances.push(first.abs_diff(*second));
    }

    println!("{}", distances.iter().sum::<u32>());
}
